backend/src/Edna/ExperimentReader/Parser.hs

```diff
@@ -0,0 +1,123 @@
+import Data.Text (replace)
+import Fmt (pretty)
+import Lens.Micro (ix, _Just)
+import Text.Read (readParen)
+
+import Edna.ExperimentReader.Error (ExperimentParsingError(..))
+import Edna.ExperimentReader.Types
+  (CellType(..), Parameter(..), ParameterType(..), PointYX(..), Signal(..), PlateUnit(..))
+import Edna.Web.Types (ExperimentalMeasurement(..))
+
+type ParserType a = Either ExperimentParsingError a
+
+-- | Based on specific CellType return its value or get and exception that cell is empty or its
+-- expected type does not satisfy actual type
+specificCellAt
+  :: Typeable a
+  => Worksheet
+  -> PointYX
+  -> CellType a
+  -> Either ExperimentParsingError a
+specificCellAt workSheet p@(PointYX position) cellType =
+  let cell = workSheet ^? ixCell position . cellValue . _Just in
+  case (cellType, cell) of
+    (_, Nothing) -> Left EmptyCell
+    (_, Just (CellText "")) -> Left EmptyCell
+
+    (CText, Just (CellText x)) -> Right x
+    (CDouble, Just (CellDouble x)) -> Right x
+
+    (CSignal, Just (CellDouble x)) -> Right $ Signal x False
+    (CSignal, Just (CellText
+      -- If signal value is in brackets then it is outlier
+      (readParen True reads . toString . replace "," "." -> ((d, "") : _))
+      )) -> Right $ Signal d True
+
+    (_, cv) -> Left $ UnexpectedCellType p cellType cv
+
+-- | Check that cell representation equals to some value
+cellSatisfy :: Worksheet -> Text -> PointYX -> Bool
+cellSatisfy workSheet value (PointYX position) = case workSheet ^? ixCell position . cellValue of
+  Just (Just x) -> pretty x == value
+  _ -> value == ""
+
+-- | Common function for splitting
+paramsSplit :: Worksheet -> ParameterType -> Int -> NonEmpty Int -> ParserType [Parameter]
+paramsSplit workSheet pType normalAxisPoint indexes = constructParams (last indexes) $
+    flip NE.filter indexes $ \c -> not $ cellSatisfy workSheet "" $ direction (normalAxisPoint, c)
+  where
+    direction :: (Int, Int) -> PointYX
+    direction points = case pType of
+      Target -> PointYX points
+      Compound -> PointYX $ swap points
+
+    -- | Construct params from their indexes
+    constructParams :: Int -> [Int] -> ParserType [Parameter]
+    constructParams _ [] = pure []
+    constructParams lastId [idx] = do
+      name <- specificCellAt workSheet (direction (normalAxisPoint, idx)) CText
+      pure [Parameter name (idx, lastId)]
+    constructParams lastId (id1 : id2 : ids) = do
+      name <- specificCellAt workSheet (direction (normalAxisPoint, id1)) CText
+      restParams <- constructParams lastId (id2 : ids)
+      pure $ Parameter name (id1, id2 - 1) : restParams
+
+parseExperimentXls :: L.ByteString -> ParserType [ExperimentalMeasurement]
+parseExperimentXls content = do
+  xlsx <- first FileParsingError $ toXlsxEither content
+  workSheet <- maybeToRight WorksheetNotFound $ xlsx ^? xlSheets . ix 0 . _2
+  processWorkSheet workSheet
+
+processWorkSheet :: Worksheet -> ParserType [ExperimentalMeasurement]
```

Comment:
We should probably have nested maps or something like that and group measurements by projects, targets… Then the size of data will be much smaller and we won't have that many layers in list monad. But the current approach is ok for now, let's revise it later.

Reply:
What do you mean by `nested maps`? \
Yes, we need to group them but I think we don't have understanding how we want to do it